Fixes and error handling rework for RAUC boot selection layer

ejoerns opened the pull request:
This series targets different issues in Barebox bootchooser (boot selection layer) implementation.

The two main ones are:

* This is one of the components that still lacks appropriate GError handling, giving us only insufficient information about failures
* barebox-state backend did not properly handle disabled slots for setting slots as primary. This also additionally requires reading state info instead of only writing it

jluebbe commented on src/bootchooser.c:
```diff
@@ -44,27 +44,36 @@ static gboolean bootchooser_order_primay(RaucSlot *slot, GString **value) {
+
+#define BOOTSTATE_PREFIX "bootstate"
+
 #if 0
```
According to the commit message, the function should be reenabled.

ejoerns replied:
This would require squashing commits if we intend to keep the commit compilable.

jluebbe replied:
Hmm, OK. Then explain that in the commit message.

jluebbe commented on src/bootchooser.c:
```diff
@@ -22,12 +22,11 @@ GQuark r_bootchooser_error_quark (void)
 	GString *order = g_string_sized_new(10);
 	GList *slots;
 
 	g_return_val_if_fail(slot, FALSE);
```
We can't return FALSE here anymore.

jluebbe commented on src/bootchooser.c:
```diff
@@ -508,9 +501,13 @@ static gboolean uboot_set_primary(RaucSlot *slot) {
-		goto out;
+	res = uboot_env_get("BOOT_ORDER", &order_current, &ierror);
+	if (!res) {
+		g_message("Unable to obtain BOOT_ORDER");
```
Maybe indicate that this is not cirtical and we fall back to a default order?

ejoerns replied:
I assume changing it to 
``` c
g_message("Unable to obtain BOOT_ORDER, using defaults");
```

would be sufficient, as the log level indicates being non-critical?

jluebbe commented on test/install-content/custom_handler.sh:
```diff
@@ -24,6 +24,14 @@ done
 exit_if_empty $RAUC_MOUNT_PREFIX
 mkdir -p $RAUC_MOUNT_PREFIX/image
 
+# This is only for testing
```
Should this change be squashed into the one changing the code?

ejoerns replied:
Yes, could, was more ore less a post-fix.

ejoerns commented:
Updated commit messages, squashed and fixed as requested.

``` diff
diff --git a/src/bootchooser.c b/src/bootchooser.c
index bf8888b..504dfa6 100644
--- a/src/bootchooser.c
+++ b/src/bootchooser.c
@@ -26,7 +26,7 @@ static GString *bootchooser_order_primay(RaucSlot *slot) {
        GString *order = g_string_sized_new(10);
        GList *slots;
 
-       g_return_val_if_fail(slot, FALSE);
+       g_return_val_if_fail(slot, NULL);
 
        g_string_append(order, slot->bootname);
 
@@ -504,7 +504,7 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
 
        res = uboot_env_get("BOOT_ORDER", &order_current, &ierror);
        if (!res) {
-               g_message("Unable to obtain BOOT_ORDER");
+               g_message("Unable to obtain BOOT_ORDER, using defaults");
                g_clear_error(&ierror);
 
                order_current = bootchooser_order_primay(slot);
```
